[CBRD-26367] (Phase-2) Optimize NUMERIC storage format

jongmin-won opened the pull request:
http://jira.cubrid.org/browse/CBRD-26367

### Purpose

- 기존(develop)에는 Fixed Numeric 타입이 고정 길이 16 byte로 저장되었으나, 이번 수정으로 precision 값에 따라 가변 길이로 저장되도록 변경됨.
- Floating Point Numeric의 경우에는 Header 2 byte를 포함하여 총 20 byte 고정 길이로 저장함.
  - Floating 타입은 입력 시점에 어떤 precision 값이 올지 알 수 없으므로, 항상 최대 크기로 저장.
  - Header : 2 byte
  - Precision = 43 (=18 byte)

### Implementation

#### 1. DB_MAX_NUMERIC_PRECISION 확장
- precision : 38 -> 43

#### 2. NUMERIC 반올림 기준 변경
- phase-1 까지는 38자리가 넘어가는 경우, 39자리에서 반올림을 수행함
- phase-2 부터 43자리가 넘어가는 경우, 44 자리에서 반올림을 수행함

#### 3. Fixed Numeric 타입의 저장 방식을 기존 고정 길이(16 byte) 에서 precision 값에 따라 가변 길이로 계산하도록 변경함. 
- precision 계산 식 : log10(256) = 2.40824
```
byte = ceil(precision / log10(256))
```
| Precision | Byte | 4-Byte Aligned |
|-----------|------|------------------|
| 1 | 1 | 4 |
| ... | ... | ... |
| 10 | 5 | 8 |
| ... | ... | ... |
| 20 | 9 | 12 |
| ... | ... | ... |
| 29 | 13 | 16 |
| ... | ... | ... |
| 38 | 16 | 16 |
| 39 | 17 | 20 |
| ... | ... | ... |
| 43 | 18 | 20 |

#### 4. 메모리 및 디스크 버퍼 관련 함수 수정
- mr_init(mem/val)_numeric
- mr_set(mem/val)_numeric
- mr_data_leng(mem/val)_numeric
- mr_data_read(mem/val)_numeric
- mr_data_write(mem/val)_numeric
- ...

### Remarks
#### 1. AUTO_INCREMENT 컬럼의 타입이 NUMERIC(1~28)로 생성된 경우, current_val 및 next_val을 읽지 못하는 문제 (해결 완료)
- 해당 문제는 수정을 완료했으며, phase-3에 반영하겠습니다.
- 아래 예시에서 Actual result의 id 값이 NULL로 출력되는 예상 원인 :
  - DB_SERIAL 카탈로그의 current_val 등의 컬럼은 NUMERIC(38) 타입으로 정의되어 있어 16바이트 단위로 값이 저장/읽기됨.
  - 반면, AUTO_INCREMENT 컬럼이 NUMERIC(1~28) 타입인 경우 4 ~12바이트만 읽기 때문에, 값이 올바르게 매칭되지 않아 NULL이 출력되는 것으로 추정됨.

```
CREATE TABLE t1 (id NUMERIC(21) AUTO_INCREMENT, name VARCHAR);
INSERT INTO t1 (name) VALUES ('test');
select * from t1;
```

- Expected result
```
  id                    name                
============================================
  1                     'test'    
```

- Actual result
```
  id                    name                
============================================
  NULL                  'test'  
```

Commit:
(Phase-2) Optimize NUMERIC storage format

ctshim commented on src/compat/dbtype_def.h:
```diff
@@ -593,7 +597,11 @@ extern "C"
 #define DB_DEFAULT_NUMERIC_DIVISION_SCALE 9
 
 /* These constants define the size of buffers within a DB_VALUE. */
-#define DB_NUMERIC_BUF_SIZE	(2*sizeof(double))
+#if 1				// used in phase-2
+#define DB_NUMERIC_BUF_SIZE	((2*sizeof(double)) + 4)
+#else				// used in phase-3
+#define DB_NUMERIC_BUF_SIZE	((2*sizeof(double)) + 2)
+#endif
```
NUMERIC_BUF_SIZE와 double 타입 2개의 크기라는 것은 어떤 관계가 있는 것인가요? 뭔가 double 두 개를 저장하는 형태에서 기인한 것인지 궁금합니다.
여기에 `+ 4` 혹은  `+ 2`하는 이유를 유추 할 수 있는 코멘트가 추가되면 좋겠습니다.

beyondykk9 replied:
굳이 이전의 코드를 반영할 필요는 없는 것 같고, 새로 계산된 수식으로 다시 정의할 필요가 있습니다.

jongmin-won replied:
@ctshim 

기존 코드의 `2 * sizeof(double)` 표기는 16바이트를 의미하기 위한 것으로 보입니다.

phase-2와 phase-3 간의 2바이트 차이는 다음과 같습니다.
- phase-2: Header 2바이트를 별도로 전달하지 않고, 전체 20바이트(데이터 포함)를 한 번에 읽고/쓰기 함.
- phase-3: Header를 별도의 2바이트로 분리하여 전달하고, 데이터 버퍼는 18바이트만 읽고/쓰기 함으로써 전체 20바이트를 사용함.

jongmin-won replied:
@beyondykk9 

최종(phase-3) `DB_NUMERIC_BUF_SIZE`의 최대 크기는 Floating NUMERIC의 전체 저장 크기인 20바이트 중, Header 2바이트를 제외한 18바이트 데이터 영역만 저장하도록 정의할 계획 입니다.

현재 phase-2에서 DB_NUMERIC_BUF_SIZE를 20바이트로 유지한 이유는 다음과 같습니다.

- 아직 Header 2바이트를 별도로 전달하지 않기 때문에, phase-3처럼 DB_NUMERIC_BUF_SIZE 값을 18바이트로 변경할 경우, 내부 연산 과정에서 precision 값이 39~43이 되는 상황에서 20바이트 읽기/쓰기 시 Overflow 문제가 발생할 수 있습니다.
- 이러한 문제를 임시로 방지하기 위해, phase-2에서는 DB_NUMERIC_BUF_SIZE를 20바이트로 설정했습니다.

참고로, 새로 계산된 수식은 각 mr_data_*_numeric() 함수 내부에서 사용되고 있습니다.

jongmin-won replied:
@beyondykk9 
오프라인에서 논의한 내용대로, `(2*sizeof(double))` 같은 표현은 제거하고 계산된 결과인 상수 값으로 직접 입력하도록 변경했습니다.

ctshim commented on src/compat/dbtype_function.i:
```diff
@@ -1501,16 +1504,52 @@ db_make_bigint (DB_VALUE * value, const DB_BIGINT num)
+db_make_fixed_numeric_internal (DB_VALUE * value, const DB_C_NUMERIC num, const int precision, const int scale,
+				const int byte_size)
+{
+  memcpy (value->data.num.d.buf + (DB_NUMERIC_BUF_SIZE - byte_size), num, byte_size);
```
`value->data.num.header` 의 정보를 fixed numeric임을 나타내도록  명확하게 지정해 주는 것이 어떤가요?

jongmin-won replied:
fixed numeric의 경우, 이미 `db_value_domain_init()` 단계에서 초기화 및 도메인 설정이 완료되며, 
이후 실제 데이터 복사를 위해 `db_make_fixed_numeric_internal()` 함수가 호출됩니다.

- 호출 순서
```
db_make_numeric() 
-> db_value_domain_init() 
-> db_make_fixed_numeric_internal()
```

따라서 `db_make_fixed_numeric_internal()` 함수 내에서
`value->data.num.header` 정보를 다시 초기화 또는 셋팅 하는 것은 중복된 작업으로 보입니다.

beyondykk9 commented on src/compat/dbtype_function.i:
```diff
@@ -90,8 +90,11 @@ STATIC_INLINE int db_make_method_error (DB_VALUE * value, const int errcode, con
 STATIC_INLINE int db_make_bigint (DB_VALUE * value, const DB_BIGINT num) __attribute__ ((ALWAYS_INLINE));
-STATIC_INLINE int db_make_numeric (DB_VALUE * value, const DB_C_NUMERIC num, const int precision, const int scale)
-  __attribute__ ((ALWAYS_INLINE));
+static void db_make_fp_numeric_internal (DB_VALUE * value, const DB_C_NUMERIC num);
```
db_make_fp_numeric_internal  보다는 db_make_float_numeric_internal 이 낫지 않을까요?

jongmin-won replied:
리뷰 감사합니다.
코드 리뷰가 모두 끝난 뒤, 머지 전에 한 번에 반영하겠습니다.

- convention 수정
`fp_numeric_*` -> `float_numeric_*`

jongmin-won replied:
convention 수정 하여 반영 했습니다.

beyondykk9 commented on src/object/object_primitive.c:
```diff
@@ -8288,8 +8288,22 @@ static void
+  else
+#endif
+    {
+      mem_length = fp_numeric_precision_to_bytes (domain->precision);
```
fp_numeric_precision_to_bytes 보다는 float_numeric_precision_to_bytes

jongmin-won replied:
리뷰 감사합니다.
코드 리뷰가 모두 끝난 뒤, 머지 전에 한 번에 반영하겠습니다.

- convention 수정
`fp_numeric_*` -> `float_numeric_*`

jongmin-won replied:
convention 수정 하여 반영 했습니다.

hyunikn commented:
Remarks 에 적힌 AUTO_INCREMENT 문제는 이번 PR 에서 해결하나요?

Commit:
1. Update the mr_data_*(val/mem) functions so that converting precision to bytes no longer performs a calculation each time, but instead uses a lookup table.
2. Replace expressions like (2 * sizeof(double)) in DB_NUMERIC_BUF_SIZE with the constant value computed from ceil(precision / log10(256)), and set that value directly.

jongmin-won commented:
@hyunikn 
AUTO_INCREMENT 관련 문제는 해결 완료됐으며, phase-3에서 해당 수정 사항을 함께 반영하겠습니다.

Commit:
Fix code style issues

hyunikn commented on src/compat/dbtype_def.h:
```diff
@@ -584,7 +584,11 @@ extern "C"
+#if 1				// used in phase-2
 #define DB_DEFAULT_NUMERIC_PRECISION 15
+#else				// used in phase-3
+#define DB_DEFAULT_NUMERIC_PRECISION DB_MAX_NUMERIC_PRECISION
```
phase 3에서 DB_MAX_NUMERIC_PRECISION (43) 이 * (임의의 precision) 을 나타내나요?

jongmin-won replied:
네, phase-3 에서 DB_MAX_NUMERIC_PRECISION(43)은 float numeric(= *) 으로 간주합니다.

hyunikn commented on src/compat/dbtype_function.i:
```diff
@@ -1501,16 +1504,52 @@ db_make_bigint (DB_VALUE * value, const DB_BIGINT num)
+ * byte_size(in):
+ */
+static void
+db_make_fixed_numeric_internal (DB_VALUE * value, const DB_C_NUMERIC num, const int precision, const int scale,
```
precision과 scale 인자는 사용되지 않고 있습니다.

jongmin-won replied:
리뷰 감사합니다.

코드 리뷰가 모두 끝난 뒤, convention 정리할 때 같이 수정하겠습니다.

jongmin-won replied:
말씀하신 부분을 반영하여 수정했습니다.

hyunikn commented on src/object/object_primitive.c:
```diff
@@ -8401,14 +8507,17 @@ mr_data_lengthmem_numeric (void *mem, TP_DOMAIN * domain, int disk)
     {
-      len = OR_NUMERIC_SIZE (domain->precision);
+      len = OR_SHORT_SIZE;
+      len += MR_NUMERIC_SIZE (domain->precision);
```
MR_NUMERIC_SIZE와 OR_NUMERIC_SIZE 의 차이는 무엇일까요. 
현재는 같은 값이던데요.

jongmin-won replied:
현재 develop 기준으로도 `OR_NUMERIC_SIZE`와 `MR_NUMERIC_SIZE`는 동일한 값으로 정의되어 있습니다.

두 매크로가 정의된 위치에 주석 내용을 보면,
NUMERIC은 `DB_NUMERIC_BUF_SIZE` 크기의 고정 길이 버퍼를 사용하고 있고,
향후 precision/scale에 따라 가변 길이로 바뀔 수 있기 때문에, 그 변화에 대비해서 각각 별도의 매크로를 둔 것으로 추측됩니다.

다만, 현 시점에서는 char 타입과 비슷하게 내부 저장은 고정 길이 버퍼를 사용하고 있고,
그 고정 길이 안에서 precision에 따라 실제 사용하는 바이트 수만 달라지는 구조로 수정 했습니다.

hyunikn commented on src/object/object_primitive.c:
```diff
@@ -8554,7 +8718,41 @@ mr_data_readval_numeric (OR_BUF * buf, DB_VALUE * value, TP_DOMAIN * domain, int
        * will copy the bits into its internal buffer.
        */
-      (void) db_make_numeric (value, (DB_C_NUMERIC) buf->ptr, domain->precision, domain->scale);
+#if 0				// used in phase-3
```
변경 내용과 관련 있는 내용은 아니지만, mr_data_readval_numeric()의 copy, copy_buf, copy_buf_len 은
사용하고 있지 않으므로 지워주면 코드 정리 차원에서 좋을 것 같습니다.

jongmin-won replied:
다른 `mr_data_readval_*` 계열 함수들도 `copy`, `copy_buf`, `copy_buf_len` 인자를 받고 있지만 실제로는 사용하지 않고 있습니다.
예: `mr_data_readval_object()` 등

해당 인자 제거는 별도 이슈로 묶어서 한 번에 정리하는 편이 더 좋을 것 같습니다.

hyunikn commented on src/query/arithmetic.c:
```diff
@@ -4544,7 +4564,19 @@ db_width_bucket_calculate_numeric (double *result, const DB_VALUE * value1, cons
+#if 1				// used in phase-2
 	      numeric_coerce_num_to_double (db_get_numeric (&n4), DB_VALUE_SCALE (&n4), &res);
+#else // used in phase-3
+	      if (DB_VALUE_PRECISION (&n4) == DB_DEFAULT_NUMERIC_PRECISION
```
DB_VALUE_PRECISION (&n4) == DB_DEFAULT_NUMERIC_PRECISION 이지만, 
DB_VALUE_SCALE (&n4) != DB_DEFAULT_NUMERIC_SCALE) 인 경우가 가능한가요. 
만약 가능하다면, 그 경우 fp numeric 이므로 else part 로 진행하는 것이 적절하지 않아 보입니다.

jongmin-won replied:
`DB_VALUE_SCALE (&n4) != DB_DEFAULT_NUMERIC_SCALE` 상황은 발생하지 않습니다.

해당 지점까지 오기 전 `sub -> sub -> div -> div` 연산이 이루어지고, 마지막 div 수행 경로는 다음과 같습니다.
```
fp_numeric_db_value_div
-> db_make_numeric
-> db_make_fp_numeric_internal
```
이 흐름에서 최종 결과는 항상 `float numeric`이며, `db_make_fp_numeric_internal()`에서 `domain.numeric_info.precision`/ `domain.numeric_info.scale`이 `DB_DEFAULT_NUMERIC_PRECISION`/ `DB_DEFAULT_NUMERIC_SCALE`로 저장됩니다.

따라서 이 경로 외에는 들어올 가능성이 없어 else 절은 필요 없을 것으로 보이지만,
width_bucket() 전체 흐름에 대한 테스트가 아직 충분하지 않아 안전 차원에서 else 조건을 추가해둔 상태입니다. 
(phase-3 리뷰 전까지 더 검증할 예정입니다.)

hyunikn replied:
> DB_VALUE_SCALE (&n4) != DB_DEFAULT_NUMERIC_SCALE 상황은 발생하지 않습니다.

그렇다면 이 조건을 검사에서 빼고 then part 에서 
```assert (DB_VALUE_SCALE (&n4) == DB_DEFAULT_NUMERIC_SCALE);``` 하는 것으로 바꾸면 좀 더 readability 가 좋지 않을까 합니다.

kangmin5505 commented on src/compat/dbtype_function.i:
```diff
@@ -90,8 +90,11 @@ STATIC_INLINE int db_make_method_error (DB_VALUE * value, const int errcode, con
-STATIC_INLINE int db_make_numeric (DB_VALUE * value, const DB_C_NUMERIC num, const int precision, const int scale)
-  __attribute__ ((ALWAYS_INLINE));
+static void db_make_fp_numeric_internal (DB_VALUE * value, const DB_C_NUMERIC num);
+static void db_make_fixed_numeric_internal (DB_VALUE * value, const DB_C_NUMERIC num, const int precision,
```
해당 파일의 함수 컨벤션으로 보이는 STATIC_INLINE을 사용 안 하시고 static 으로 선언하신 이유가 있을까요?

kangmin5505 replied:
[완료]
다른 함수는 external 로 선언되어 있는데, 해당 함수는 내부에서만 쓰이는 것이기 때문에 static으로 선언하셨다고,
구두로 설명 들었습니다.

kangmin5505 commented on src/object/object_primitive.c:
```diff
@@ -8313,20 +8328,69 @@ mr_setmem_numeric (void *mem, TP_DOMAIN * domain, DB_VALUE * value)
+
+	      if (value->data.num.header.scale < 0)
+		{
+		  header[0] = DB_VALUE_NUMERIC_HEADER_PRECISION (value) | 0x80;
```
0x80 이 여러 곳에서 사용이 되는데, 같은 의미로 사용되는 거라면 메크로로 선언하는 것이 의미 표현과 유지보수에 좋을 거 같습니다.

jongmin-won replied:
리뷰 감사합니다. 

numeric value에서 사용하던 부호 비트(0x80)는 `NUMERIC_VALUE_SIGN_BIT_MASK`로,
header에서 음수 scale을 표시하기 위해 사용하던 부호 비트(0x80)는 `NUMERIC_HEADER_SCALE_SIGN_BIT_MASK`로 각각 대체하여 수정했습니다.

hyunikn commented on src/query/query_opfunc.c:
```diff
@@ -6164,7 +6164,22 @@ qdata_unary_minus_dbval (DB_VALUE * result_p, DB_VALUE * dbval_p)
+      db_make_numeric (result_p, db_get_numeric (dbval_p), DB_VALUE_PRECISION (dbval_p), DB_VALUE_SCALE (dbval_p),
+		       DB_NUMERIC_BUF_SIZE, false);
+#else // used in phase-3
+      if (DB_VALUE_PRECISION (dbval_p) == DB_DEFAULT_NUMERIC_PRECISION)
```
phase-3 에서 "DB_VALUE_PRECISION (dbval_p) == DB_DEFAULT_NUMERIC_PRECISION " 여부에 따라 분기하는 코드가 db_make_numeric 내부로 들어갈 수 있을 것 같습니다. 
```
db_make_numeric(result_p, dbval_p, DB_NUMERIC_BUF_SIZE, false);
```
이렇게 쓰는 거죠.

Commit:
1. Replaced the hard-coded 0x80 used in the numeric value and header with dedicated macros (NUMERIC_VALUE_SIGN_BIT_MASK and NUMERIC_HEADER_SCALE_SIGN_BIT_MASK) to improve readability.
2. Removed an unused parameter from the db_make_fixed_numeric_internal() function.

Commit:
Removed an unused parameter from the db_make_fixed_numeric_internal() function.

Commit:
Updated the naming convention by renaming fp_numeric_* functions to the float_numeric_* prefix.

Commit:
Merge remote-tracking branch 'upstream/feature/scale_range_fp_numeric' into CBRD-26367

Commit:
Merge remote-tracking branch 'upstream/feature/scale_range_fp_numeric' into CBRD-26367
